### ISSUE TITLE
Improve dynamic completions

### ISSUE THREAD

**Issue by casey** [casey/just#2406]
Just previously used clap's statically generated completion scripts. These scripts would complete flags and options, but wouldn't complete recipes, since that requires invoking just itself to find and parse the justfile. To remedy this, we started patching the generated completion scripts.

This slowly turned into a maintainance nightmare, as the patches grew in number and complexity, and had to be manually adapted every time the base completion scripts changed.

Clap has a new dynamic completion engine, tracked in https://github.com/clap-rs/clap/issues/3166.

In #3167 I switched to the new completion engine. Strictly speaking, this was a regression, since we lost the ability to complete recipes, but I didn't want to try to do everything in one PR, so I thought it was better to merge it and add back the missing functionally in follow-up PRs.

To do:
- [ ] Bash splits completion arguments on `:`, which breaks completion of `::`-separate recipe paths. We should upstream a fix to `clap_complete`, since it's basically never correct to split on `:`. I have a draft PR open for this: https://github.com/clap-rs/clap/pull/6311
- [x] Complete arguments with variable names (suffixed with `=`, recipes, and file paths (since they could be recipe arguments)
- [ ] Complete arguments more intelligently. I'm not sure how much we can do here, but it would be ideal to only complete overrides if we haven't gotten a recipe name or an argument, and to only complete recipes when we're in a position where we're expecting a recipe (i.e., we're not in position where we're completing a recipe argument). This may be tough. We don't know where exactly in the positional arguments we're completing, so we don't necessarily know what we're completing. However, this enables a bunch of stuff, like showing the parameter name when completing a recipe parameter, or completing flags and options.
- [x] Complete variable names for `--set`
- [ ] Support proper completions for nushell
- [ ] Don't complete variable names in value position for `--set`, i.e., `--set foo <tab>`
- [x] Complete recipes for `--show`
- [ ] Look through all arguments for opportunities to improve completions (e.g., only complete files or directories, or add a new custom completer)
- [x] Add instructions for each supported shell for how to load shell completion scripts, including how to lazy load them
- [x] Close all PRs and issues related to the old completion scripts
- [ ] The patched zsh completion script had a nice feature where if pressed tab after completing a recipe, it would show you the recipe as a hint message. This was nice, we should try to replicate it.
- [ ] When completing recipes, should we show the docs, the parameters, or both?
- [ ] When completing a variable for an override, we should complete `VAR=` and leave the cursor after the `=` (`-qS =` in zsh) so the user doesn't have to backspace.
- [ ] Complete paths with `~`, `~USER`, `$VARIABLE` (maybe not wise, since variable references are different across shells), and anything else? (I think this is only broken in zsh)
- [x] Complete group names for `--group`
- [ ] Complete modules for `--list` (seems to be a bug in clap)
- [x] Complete recipe path for `--usage`
- [ ] `--chooser`, `--cygpath`, `--shell`, `--command` (first only) complete files and binary names in PATH. (this is maybe crazy, since there might be many binaries on PATH)
- [ ] `--dotenv-filename` Maybe we should complete files? It's not technically correct, since they should be filenames only, but it's probably better than not completing files.
- [ ] Report back to clap with any feedback on integration (felt so good to nuke the old scripts, huge improvement in functionality and consistency, I think maybe I would make it more clear in documentation that the feature is unstable but very usable since I probably waited too long to try it, we are parsing args, and we rely on being able to remove the arg prefix, and any kind of testing relies on setting the argument correctly to call the binary, bash alias completion depends on knowing the name of the completion function)
- [ ] Figure out a way to print errors in completion engine 
- [ ] Handle space-separated paths in arguments and `--show`
- [ ] Reconsider including recipe doc comments in completions, it makes them take up much more space
- [ ] Autocomplete fallback recipes (filter recipes can cannot be executed via fallback)

**Comment by heyrict** [comment]
After investigations into the experimental clap completion engine, I think it is a promising yet quite simplified implementation, at least for zsh. For example, it [saves all completion variables in a array](https://github.com/clap-rs/clap/blob/f6a6701e60f372c2104c6aca413c86b996e3e026/clap_complete/src/env/shells.rs#L382-L397), split in two by its contents and put all of them in `_describe 'values'` in a single function, while [complex commands (e.g. systemctl)](https://github.com/systemd/systemd/blob/main/shell-completion/zsh/_systemctl.in#L113) frequently use shell-specific features to make the completion results clean and user-friendly.

In this architecture, it would be quite straight-forward if we have pre-defined subcommands and arguments. However, in order to complete flexibly with shell-specific functions (e.g. `_message` for displaying help messages, `_describe -o/-t` for controlling completion behavior), we have to either improve the *generic* clap completion engine upstream (difficult, error-prone and time-consuming) or fork one specifically for us (another mainteinance hell).

I definitely agree that patching an auto-generated completion script would not be sustainable in the long run. I think providing a completion function manually, or maybe `format!` one based on `enum Subcommand` would be a better choice for us.

**Comment by laurigates** [comment]
### Idea: group recipe completion candidates by `[group(...)]`

A small addition for the tracking list, plus a report-back from the clap side.

**User value:** `just <TAB>` could render recipes grouped under their
`[group('...')]` headings instead of one flat list — so a justfile with
`[group('build')]` / `[group('test')]` recipes completes as labelled,
switchable sections. In zsh this also lets each group be styled/colorized
independently (e.g. via fzf-tab), which makes large justfiles much easier to
scan at the prompt.

**Where it would hook in:** `candidate_recipes()` in `src/completer.rs` already
builds each `CompletionCandidate::new(path).help(...)`. The group data is right
there too — `Recipe::groups()` (`src/recipe.rs`) exposes a recipe's groups at
exactly that point — so the candidate just needs the group attached as a
display tag.

**Blocker (the report-back):** `CompletionCandidate` today carries only `value`
+ `help()`, with no group/tag field. Grouping needs a `CompletionCandidate`
tag API, which is in clap PR clap-rs/clap#6334 ("group options by tag (in
zsh)", from clap-rs/clap#6320) — currently open, not merged. Once it lands it
also means bumping the `clap_complete = "=4.6.5"` pin. It's zsh-first: bash,
fish, and PowerShell don't have an equivalent tag-grouping mechanism yet.

**Evidence the UX is worth it:** I have a working downstream proof-of-concept —
a zsh completer that parses `just --dump --dump-format json` and groups recipes
via `_describe` headers, rendered with fzf-tab. It confirms the grouped list is
a clear improvement over the flat one; the only thing keeping it out of `just`
proper is the missing clap tag API above.

Filing here rather than as a new issue, since #2406 already invites reporting
clap-integration feedback. Happy to share the PoC completer if useful.
